File: 3/go/part2/part2.go
package main

import (
	"log"
	"os"
	"regexp"
	"strconv"
)

func main() {
	if len(os.Args) < 2 {
		log.Fatal("provide part and input file")
	}
	inputFile := os.Args[1]

	input, err := os.ReadFile(inputFile)
	if err != nil {
		log.Fatal(err)
	}

	c, _ := regexp.Compile(`mul\([0-9]+,[0-9]+\)|do\(\)|don't\(\)`)
	d, _ := regexp.Compile(`[0-9]+`)

	commands := c.FindAllString(string(input), -1)
	enabled := true

	total := 0

	for _, cmd := range commands {
		log.Println("command ", cmd)
		switch cmd {
		case "do()":
			enabled = true
		case "don't()":
			enabled = false
		default:
			if !enabled {
				continue
			}
			digits := d.FindAllString(string(cmd), -1)
			first, err := strconv.Atoi(digits[0])
			if err != nil {
				log.Fatal(err)
			}
			second, err := strconv.Atoi(digits[1])
			if err != nil {
				log.Fatal(err)
			}
			total += first * second
		}
	}
	log.Println(total)
}

File: 3/py/part2.py
import sys
import string

if len(sys.argv) < 2:
    print("provide input file")

inputfile = sys.argv[1]
input = ""

with open(inputfile, "r") as f:
    input = f.read()


def parseMul(s: string) -> int:
    if len(s)<5 or s[0] != "(": return 0
    first = ""
    second = ""
    comma_found = False
    for c in s[1:]:
        if c in string.digits:
            if not comma_found:
                first += c
            else:
                second += c
        elif c == "," and not comma_found:
            comma_found = True
        elif c == ")":
            try:
                return int(first)*int(second)
            except Exception as e:
                print(e)
                break
        else:
            break
    return 0

input = input.split("don't()")
enabled = True
total = 0

for i in input:
    dos = i.split("do()")
    if not enabled:
        if len(dos) == 0:
            continue
        dos = dos[1:]
    for d in dos:
        muls = d.split("mul")
        for m in muls:
            total += parseMul(m)

    enabled = False

print(total)